README.md
# 特别说明：本库仅提供企业内部服务，不向外部提供商业性的数据服务和支持

dbpystream API 主要为已经购买了数据供应商API服务的企业内部提供向自己员工内部的数据服务的API接口，这种服务是一种API的镜像服务。

## dbpystream API的目的

对于不少企业来说，购买了很多不同类的数据，比如wind,同花顺，joinquant,财汇，聚源等等。很多是以账号的形式提供服务，也有部分是以落地数据库的形式。但这些形式，更多的只是提供给企业内部的局部人员使用，在数据共享和数据探索，以及服务上，还是处于一个相对窄的面上。对企业来讲，这些数据并没有得到全面的使用，也没有把数据信息资产化，并且流动起来，让业务产生更大的价值。

基于上面的痛点 ，dbpystream的目的是根据自身的业务需要，把企业内部的数据进行服务化，让企业内的员工能方便得到数据服务，而不再让数据服务只是一小部分人的特权，让数据可以在更广泛的层面得到充分的利用，并发挥数据驱动业务的价值。

从更长远的角度，数据的广泛利用，有利于在公司、行业层面发挥价值，同时也让大家更加重视数据的内在价值，也有利提升行业的生态环境。

## dbpystream API 的特点

1、函数基本一致。在接口规范上，尽量保持和原生的API接口相近，尽量让大家上手比较快,个别上函数签名和返回值上可能会有微调；
2、基于内部使用。 API的服务范围是基于内部的需要开发的实用接口部分，其要比原生API范围的小；
3、数据上“源汁源味"。数据上不对源数据做相应的清洗和加工。

## dbpystream API 与其它数据源API不同

1、定位不同。dbpystream是对内服务。

和供应商提供的API服务不同，dbpystream是对内的，更相对于是一个局域网。数据源API是对外赋能，dbpystream是对内赋能，而不是对外商业化。

2、上下游关系。dbpystream依赖于供应商的API服务。

供应商的API的服务质量是dbpystream的天花板，没有优秀的供应商服务（源头），就不可能有好的内部服务。

3、共生关系。dbpystream只做整合部分，不能替代供应商API，是对供应商API服务包装，为了让服务更好的落地，而供应商API做的是一套完整的生产和质量管理流程。

## dbpystream API展望

目前，data api 市场和服务还是较初级的阶段，感谢市场上为数不多的数据提供商，可以让大家有了接触api服务的可能。在此基础上，dbpystream api才有了二次的可能，在此特别致谢，joinquant, tushare、同花顺等数据服务厂商，也感谢为市场提供免费数据的sina财经等网站。

目前，dbpystream主要的数据服务商有：jqdatasdk, 未来计划支持更多。

dbpystream只对企业内部提供数据服务,也期待能与数据供应商一起为整个数据生态而努力！

## 安装方法

```python

    pip install dbpystream

```

## 实例说明

以下jqdatasdk为例，本库提供了jqdatasdk类似的支持，主要函数签名及参数基本上与jqdatasdk一致

其中，比如，你希望通过get_price获取历史1 min bar行情，首先，你需要在首次登陆时需要使用到auth函数。

auth:

```python
    def auth(username=None,password=None) -> None
```

说明：
    可以选择手动同时填入两个参数，也可以使用从环境变量中进行设置。默认是从环境变量中;如果从环境变量设置相关变量，建议以"dbpystream_username"，"dbpystream_password"来命名。
    当前面username和password参数必须显示指定和输入。这两个参数，需要由内部分发。

在完成登陆后，你就可以直接调用get_price函数了。

get_price:

```python
    def get_price(security,start_date,end_date,frequency,fq='pre',fields=None) -> pd.DataFrame:
```

关于get_price的用法：

```python
    其中主要参数：
    1、security 
        一支股票代码或者一个股票代码的list。股票代码或期货合约要包含交易后缀，比如.XSHE,XSHG等等。
    2、start_date
        与 count 二选一，不可同时使用. 字符串或者 datetime.datetime/datetime.date 对象, 开始时间.
    3、end_date
        格式同上, 结束时间, 默认是'2015-12-31', 包含此日期. 注意: 当取分钟数据时, 如果 end_date 只有日期, 则日内时间等同于   00:00:00, 所以返回的数据是不包括 end_date 这一天的.
    4、frequency
        'daily'(等同于'1d'), 'minute'(等同于'1m'),  默认值是daily
    5、fields：有3个选项:，
        (1) None :
            只支持['code','datetime','open', 'close', 'high', 'low', 'volume', 'money']
        (2)'all' : 即全选，
            是指[code","datetime","open","close","low","high","volume","money","factor","high_limit","low_limit","avg","pre_close","paused","open_interest"]
        (3)选取字段: 比如["datetime","open","close"]
            是指从"all"对应的字段全集中，选取所需要的参数组合。
    6、fq:，是指复权参数，其中,"pre"指前复权，"post"指后复权，None为不复权。
        共有'pre',"post",None三个选项；其中默认值为，"pre"。

    如,get_price("000005.XSHE", "2022-01-01", "2022-10-05", "minute", "post")
  
                datetime   open  close   high    low   volume      money
    1 2022-01-04 09:31:00  23.51  23.51  23.51  23.41  54328.0  1276801.0
    2 2022-01-04 09:32:00  23.51  23.41  23.51  23.41   6730.0   157603.0
    3 2022-01-04 09:33:00  23.41  23.41  23.41  23.41  13704.0   320829.0
    4 2022-01-04 09:34:00  23.41  23.41  23.41  23.41   5507.0   128928.0
```

## 整体调用参考代码

```python

    import pandas as pd 
    import dbpystream
    import dbpystream.api as db

    start_date = "2020-09-30" # 日期字符串格式
    end_date   = "2022-11-01" # 日期字符串格式
    frequency  = "minute" # daily
    fq         = "post" # pre:前复权[默认]，None
    db.auth() # 第一次调用，需要登陆一下
    codes = ["159707.XSHE"] #603619.XSHG,ZN2210.XSGE
    for security in codes:
        df  = db.get_price(security,start_date,end_date,frequency,fq) ## data为返回的是pd.dataframe格式的数据；
        print(f"code : {security} df shape : {df.shape} ")
        print(f"{df.head()}")

```

输出：

```python
code : 159707.XSHE df shape : (52320, 8)  
          code            datetime   open  close   high    low      volume       money
0  159707.XSHE 2021-11-12 09:31:00  1.010  1.008  1.010  1.007  22446600.0  22635524.0
1  159707.XSHE 2021-11-12 09:32:00  1.008  1.004  1.008  1.001   2566100.0   2574806.0
2  159707.XSHE 2021-11-12 09:33:00  1.005  1.007  1.008  1.005   5207000.0   5239605.0
3  159707.XSHE 2021-11-12 09:34:00  1.007  1.008  1.009  1.007   1583300.0   1596114.0
4  159707.XSHE 2021-11-12 09:35:00  1.005  1.006  1.007  1.005   4935000.0   4966423.0
```

这样，你就可以开始使用dbpystream了。

如果有任何问题，欢迎邮件至rustroom@163.com。


CHANGE_LOG.md
# CHANGE_LOG.md

## 2022年11月 19日： 0.51版本发布


test/test.py

import pandas as pd 

import time as t
import os
import sys
import sys,os
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
import dbpystream
import dbpystream.api as db

def test_get_price():
    start_date = "2022-9-10" # 日期字符串格式
    end_date   = "2023-01-30" # 日期字符串格式
    frequency  = "minute" # daily
    fq         = "pre"#"pre"#"post" # pre:前复权[默认]，None
    #codes = ["IF9999.CCFX","TF9999.CCFX"] #603619.XSHG,ZN2210.
    #codes = ["IF9999.CCFX","TF9999.CCFX","000001.XSHE"] #603619.XSHG,ZN2210.
    codes = "600036.XSHG"

    flds  = ["datetime","close"]
    t0 = t.time()

    df  = db.get_price(codes,start_date,end_date,frequency,fq) ## data为返回的是pd.dataframe格式的数据；
    print(f"code : {codes} df shape : {df.shape} ")
    print(f"-------------{fq}--------------------")
    print(f"{df.head()}")
    print(f"{df.tail()}")
    print(f"--------------{fq}-------------------")
    print(f"cost time :{t.time()-t0} seconds!")


def test_get_all_securities():
    _type ="stock"
    _date ="2022-10-11"
    df = db.get_all_securities(types=[_type], date=_date) 
    print(df.head())

def test_get_index_weights():
    _date ="2011-05-31"
    index_id = "000001.XSHG"
    df = db.get_index_weights(index_id, _date) 
    print(df.head())

def test_get_industry_stocks():
    _date ="2015-05-31"
    industry_id = "I64"
    df = db.get_industry_stocks(industry_id, _date) 
    print(df.head())

def test_get_industry():
    _date ="2005-06-01"
    code = "600519.XSHG"
    df = db.get_industry(code, _date) 
    print(df.head())

def test_get_trade_days():
    start_date ="2021-01-01"
    end_date ="2022-10-01"
    df = db.get_trade_days(start_date, end_date)
    print(f"get_trade_days : {df}")

if __name__== "__main__" :
    #db.auth("178*******","6******") # 可以明文输入，可也以在环境变量中设置
    db.auth() #从环境变量设置相关变量，建议以"dbpystream_username"，"dbpystream_password"来命名
    test_get_price()
    test_get_all_securities()
    test_get_index_weights()
    test_get_industry_stocks()
    test_get_trade_days()
    test_get_industry()





dbpystream/utils.py
import uuid
import sys 
import pandas as pd
import pickle
import pyzstd
import os
from os import getenv
import dbpystream.method as method
import time as t

def get_username_from_env():
    value  = None
    for name in ["username", "user", "account", "mob"]:
        for prefix in ["dbpystream","dbpystreamapi","dbpystream_api","dbpystream_sdk"]:
            value = getenv('_'.join([prefix, name]).upper())
            if value:
                return value 
    return value        

def get_password_from_env():
    value =  None
    for name in ["password", "passwd","pwd"]:
        for prefix in ["dbpystream","dbpystreamapi","dbpystream_api","dbpystream_sdk"]:
            value = getenv('_'.join([prefix, name]).upper())
            if value:
                return value 
    return None
def get_account_info_from_env() -> dict:

    username = get_username_from_env()
    password = get_password_from_env()
    _dict = {"username":username,"password": password}
    return _dict

def get_mac_address() -> str:
    mac = uuid.UUID(int=uuid.getnode()).hex[-12:].upper()
    return f'{mac[0:2]}-{mac[2:4]}-{mac[4:6]}-{ mac[6:8]}-{mac[8:10]}-{mac[10:]}'

def parse_response_data(raw_data,decompress_mode ="zstd") -> pd.DataFrame :
    t0 = t.time()
    df = pd.DataFrame()
    #print(f"raw_data -> size : {sys.getsizeof(raw_data)} bytes! ")

    if not raw_data or isinstance(raw_data,str):
        print(f"response raw_data : {raw_data} is str! 或为空！")
        return df
    if decompress_mode =="zstd":
         _msg = pyzstd.decompress(raw_data)
    else :
        _msg = raw_data

    #print(f"decompress _msg -> size : {sys.getsizeof(_msg)} bytes! ")

    if raw_data :
        decompress_data = pickle.loads(_msg) # dict
        assert isinstance(decompress_data,dict),"parse_response_data解压后数据格式不是dict类型！"
        t1 = t.time()
        #print(f"parse_response_data [get_price] cost time：{t1-t0} ")    
        df   = pd.DataFrame(decompress_data)
    return df

   
def get_urls_from_methods(host) -> dict:
    _methods = method.get_methods()
    _urls = {}
    for _method in _methods:
        assert "url" in _methods[_method], "method.py文件中的url字段设置请确认是否正确!"
        url = host +_methods[_method]['url']
        _urls[_method] = url
    return _urls

def get_fields(cls):
    return  dir(cls)

def check_code(security_code) -> bool:

    if "." in security_code:
        code,exchange = security_code.split(".",1) 
        print(f"code : {code} exchange : {exchange}")
        if exchange not in ["XSHG","XSHE","CCFX","XDCE","XSGE","XZCE","XINE"]:

            return False
        else:
            if exchange in ["XSHG","XSHE"]:#股票交易所，沪深交易所

                return code.isdigit()     # str.isdigit() 判断所有字符都是数字
            elif exchange in ["CCFX","XDCE","XSGE","XZCE","XINE"]: #期货交易所

                return code.isalnum()     # str.isalnum() 判断所有字符是字母或数字两种构成
            else:
                return False # 北交所，新三板之类，债券市场
    else:

        return security_code.isalnum()    


def is_valid_date(date) -> bool:
    try:
        if ":" in date:
            t.strptime(date, "%Y-%m-%d %H:%M:%S")
        else:
            t.strptime(date, "%Y-%m-%d")
        return True
    except:
        return False

def check_get_price_params(params) ->bool:
    '''
    data : 
    {
        "security"  : "000005.XSHE", 
        "start_date": "2022-01-01", 
        "end_date"  : "2022-10-05", 
        "frequency" : "minute", 
        "fields"    : []
        "fq"        : "post","pre",None
    }
    '''
    is_all_valid = True 
    fields = ["code","datetime","open","close","low","high","volume","money","factor","high_limit","low_limit","avg","pre_close","paused","open_interest"]
    if not isinstance(params,dict):
        print("--------------------error-----------------------\n")
        print(f"params : {params} 不是dict类型！")
        print("--------------------error-----------------------\n")
        is_all_valid = False
    
    if  "start_date" in params and not is_valid_date(params["start_date"]):
        print("--------------------error-----------------------\n")
        print(f"start_date : {params['start_date']} 格式不对！")
        print("--------------------error-----------------------\n")
        is_all_valid = False
        
    if "end_date" in params and not  is_valid_date(params["end_date"]):
        print("--------------------error-----------------------\n")
        print(f"end_date : {params['end_date']} 格式不对！")
        print("--------------------error-----------------------\n")
        is_all_valid= False
    if "security" in params and isinstance(params["security"],str) and not check_code(params["security"]):
        print("--------------------error-----------------------\n")
        print(f"security : {params['security']} 请确认是否正确！")
        print("--------------------error-----------------------\n")
        is_all_valid= False
    if "frequency" in params and params["frequency"] not in ["minute","daily"]:
        print("--------------------error-----------------------\n")
        print(f"frequency : {params['frequency']} 请确认是否正确！")
        print("--------------------error-----------------------\n")
        is_all_valid= False
    # fields
    if "fields" in params and params["fields"] :
        for field in params["fields"] :
            if  field not in fields:
                print("--------------------error-----------------------\n")
                print(f"fields中字段 : {field} 不在指定fields {fields} 内！")
                print("--------------------error-----------------------\n")
                is_all_valid= False
                break
    
    if "fq" in params :
        if params["fq"] not in ["pre","post",None]:
            print("--------------------error-----------------------\n")
            print(f"fq 复权参数: {params['fq'] }!")
            print("--------------------error-----------------------\n")
            is_all_valid= False
    return is_all_valid

def check_all_securities_params(params)-> bool:
    """_summary_
      type: stock
      date: date
    """
    if "types" not in params :
        return False
    else:
        for _type in params["types"]:
            if _type not in ["stock","futures","etf","index"]:
                return False
    if "date" in params and not is_valid_date(params["date"]):
        return False
    return True

def check_index_weights_params(params) -> bool:
    """_summary_
      index_code: code
      date: date
    """
    print(f"params : {params}")
    if "index_id" not in params :
        return False

    if "date" in params and not is_valid_date(params["date"]):
        return False
    
    return True

def check_industry_stock_params(params) ->bool:
    """_summary_
      industry_code: code
      date: date
    """
    print(f"params : {params}")
    if "industry_code" not in params :
        return False

    if "date" in params and not is_valid_date(params["date"]):
        return False
    return True

def check_industry_params(params) -> bool:
    if "security" not in params :
        return False

    if "date" in params and not is_valid_date(params["date"]):
        return False
    return True

def check_trade_days_params(params) ->bool:

    if "start_date" in params and not is_valid_date(params["start_date"]):
        return False
    
    if "end_date" in params and not is_valid_date(params["start_date"]):
            return False
    return True

def check_all_trade_days_params() ->bool:
    return True
def check_security_info_params(params) ->bool:

    if "code" not in params or not check_code(params["code"]):
        return False
    return True

def check_index_stocks_params(params) ->bool:
    
    if "index_id" not in params :
            return False

    if "date" in params and not is_valid_date(params["date"]):
        return False
    return True

def check_method_params(params,method_name) -> bool : 
    #注：部分函数还没上线，会返回False
    if method_name == "get_price":
        return check_get_price_params(params)
    elif method_name == "get_all_securities":
        return check_all_securities_params(params)
    elif method_name =="get_index_weights":
        return check_index_weights_params(params)
    elif method_name == "get_industry":
        return check_industry_params(params)
    elif method_name == "get_industry_stocks":
        return check_industry_stock_params(params)
    elif method_name == "get_trade_days":
        return check_trade_days_params(params)
    elif method_name =="get_all_trade_days":
        return True
    elif method_name =="get_security_info":
        return check_security_info_params(params)
    elif method_name =="get_index_stocks":
        return check_index_stocks_params(params)
    else:
        return False
    


dbpystream/client.py
import requests
import time as t
import json
import pandas as pd

import itertools
import random
import toml
import sys
import dbpystream
import dbpystream.utils as utils
import threading
from requests.exceptions import RequestException


class LDClient(object):

    _threading_local = threading.local() 
    _auth_params = {}  
    _token = None
    _host  = "http://47.122.40.16" 
    _port  = 8080 
    _urls = utils.get_urls_from_methods(_host)
    _request_id_generator = itertools.count(random.choice(range(0, 100000, 100)))
    
    def __init__(self,username,password) -> None:

        self.client = None
        self.username = username
        self.password = password
        

    def get_login_header(self) -> dict:
        return {"Content-Type":"application/json"}

    def get_query_headers(self) -> dict:
        if self._token:
            headers = {"Content-Type": "application/json",
                "Authorization": self._token,
                "mac": utils.get_mac_address(),
                "request_id": str(next(self._request_id_generator))
                }
            return headers
        else:
            raise Exception("token:为空，请确认账户和密码是否正确 或登陆是否正常！")
    
   
    def get_method_url(self,method):
        assert method in self._urls,f"{method} 不在{self._urls}中！"
        return self._urls[method]

       
    def get_login_params(self) -> bytes:
       mac = utils.get_mac_address()
       request_id = str(next(self._request_id_generator))
       params = {"username":self._auth_params["username"],"password":self._auth_params["password"],"mac":mac,"request_id":request_id}
       js_params = json.dumps(params)
       return js_params

    def get_token(self) -> str:
        url = self.get_method_url("auth")
        #print(f" url -> : {url}")
        headers = self.get_login_header()
        _params = self.get_login_params()
        token =""
        try:
            response = requests.post(url,data = _params,headers = headers)# params 
        except RequestException as e:
            raise Exception(f"get_token -> 请求出现异常，具体原因：{str(e)}")
        except Exception as e:  
            raise Exception(f"get_token -> 系统出现异常，具体原因 :{str(e)}")
        else:
            if response.status_code == 200:
                token = response.text
            else:
                _str_msg = f'get_token -> 请求错误：{str(response.status_code)} 原因：{str(response.reason)}'
                raise Exception(_str_msg)
        return token
    
    
    @classmethod
    def auth(cls,**params) : 
        cls.set_auth_params(**params)
        token = cls.instance().get_token()
        if token :
            cls.set_token(token)
            print("登陆成功！")
        else:
            raise Exception("登陆失败！请确认账户和密码是否正确！")
    
    @classmethod
    def set_token(cls,token) -> None:
        if token:
            cls._token = token

    @classmethod
    def set_auth_params(cls, **params):
        if params != cls._auth_params and cls.instance():
            cls._threading_local._instance = None
        #login(username=None,password=None,from_env =True) -> None: 
        if params["username"] and params["password"]: # 即有账户，又有密码
            cls._auth_params = params 
        else:
            params = utils.get_account_info_from_env()
            assert params["username"] and  params["password"],"从环境变量中没有取到相应的账户和密码！请确认是否已设置正确！或在auth函数中输入账户和密码！"
            cls._auth_params = params
    
    @classmethod
    def instance(cls):
        _instance = getattr(cls._threading_local, '_instance', None)
        if _instance is None:
            if not cls._auth_params:
                params =  utils.get_account_info_from_env()
                cls._auth_params = params
              
            if cls._auth_params:
                _instance = LDClient(**cls._auth_params) ##初始化
                cls._threading_local._instance = _instance
        return _instance

    def __call__(self, method, **kwargs):

        return self.query(method, kwargs)
        print('query ok!')

    def  __getattr__(self, method):
        return lambda **kwargs: self(method, **kwargs)
       
    def query(self,method_name,params) -> pd.DataFrame:
        url = self.get_method_url(method_name)
        #print(f"query -> url :  {url} ")
        headers = self.get_query_headers()
        data = json.dumps(params) ## 不能少
        #print(f"query -> data : {data} method_name: {method_name}")
        
        assert utils.check_method_params(params,method_name),"请确认该函数是否上线、或输入参数顺序或参数设置是否正确！"
        #print(f"query     :  -> token : {self.token} auth_params :{self._auth_params}")
        #print(f"headers   :  {headers}")
        
        df = pd.DataFrame()
        try:
            response = requests.post(url,data = data,headers = headers) 
        except RequestException as e:
            print(f"请求出现异常，具体原因：{str(e)}")
        except Exception as e:  
            print("系统出现异常，具体原因 :{str(e)}")
        else:
            if response.status_code == 200:
                result = response.content
                df = utils.parse_response_data(result)
            else:
                print(f'请求错误：{str(response.status_code)} 原因：{str(response.reason)}')
        return df
 










dbpystream/api.py

import sys
import requests
import time as t
import pandas as pd

import dbpystream
from  dbpystream.client import LDClient

# 如果username和password都不选，表示会从环境变量中来取相关的账户和密码
def auth(username=None,password=None) -> None: 
    return LDClient.instance().auth(**locals())

# 获取历史数据，可查询多个标的多个数据字段，返回数据格式为 DataFrame
def get_price(security,start_date,end_date,frequency,fq='pre',fields=None) -> pd.DataFrame:
    return LDClient.instance().get_price(**locals())

    '''
    security: 一支股票代码或者一个股票代码的list
    start_date: 与 count 二选一，不可同时使用. 字符串或者 datetime.datetime/datetime.date 对象, 开始时间.
    end_date: 格式同上, 结束时间, 默认是'2015-12-31', 包含此日期. 注意: 当取分钟数据时, 如果 end_date 只有日期, 则日内时间等同于 00:00:00, 所以返回的数据是不包括 end_date 这一天的.
    frequency: 'daily'(等同于'1d'), 'minute'(等同于'1m'),  默认值是daily
    fields: 
     None  :只支持['code','datetime','open', 'close', 'high', 'low', 'volume', 'money']
     'all' : 全选，即[code","datetime","open","close","low","high","volume","money","factor","high_limit","low_limit","avg","pre_close","paused","open_interest"]
     list: all全部字段中选取部分字段，比如
    fq: 'pre',"post",None,默认值，"pre"

    get_price(["000005.XSHE"], "2022-01-01", "2022-10-05", "minute", "post")
    
                datetime   open  close   high    low   volume      money
    1 2022-01-04 09:31:00  23.51  23.51  23.51  23.41  54328.0  1276801.0
    2 2022-01-04 09:32:00  23.51  23.41  23.51  23.41   6730.0   157603.0
    3 2022-01-04 09:33:00  23.41  23.41  23.41  23.41  13704.0   320829.0
    4 2022-01-04 09:34:00  23.41  23.41  23.41  23.41   5507.0   128928.0

    '''

# 获取指定日期区间内的限售解禁数据
def get_locked_shares(stock_list, start_date, end_date, forward_count):
    return LDClient.instance().get_locked_shares(**locals())
    '''
    获取指定日期区间内的限售解禁数据

    参数

        stock_list: 一个股票代码的 list
        start_date: 开始日期
        end_date: 结束日期
        forward_count: 交易日数量， 可以与 start_date 同时使用， 表示获取 start_date 到 forward_count 个交易日区间的数据

    返回值

        pandas.DataFrame， 各 column 的含义如下:
            day: 解禁日期
            code: 股票代码
            num: 解禁股数
            rate1: 解禁股数/总股本
            rate2: 解禁股数/总流通股本

    示例

    # 在策略中获取个股未来500天的解禁情况
    get_locked_shares(stock_list=['000001.XSHE', '000002.XSHE'], start_date=context.current_dt, forward_count=500)
    '''

#获取指数成份股
def get_index_stocks(index_symbol, date=None)->list:
    return LDClient.instance().get_index_stocks(**locals())
    
    '''
    get_index_stocks(index_symbol, date=None)
    参数
    - index_symbol, 指数代码
    - date: 查询日期, 一个字符串(格式类似’2015-10-15’)或者[datetime.date]/[datetime.datetime]对象, 可以是None, 使用默认日期. 这个默认日…
    '''

#获取指数成分股权重
def get_index_weights(index_id, date=None) ->pd.DataFrame:
    return LDClient.instance().get_index_weights(**locals())
    '''
    参数

        index_id: 代表指数的标准形式代码， 形式：指数代码.交易所代码，例如"000001.XSHG"。
        date: 查询权重信息的日期，形式："%Y-%m-%d"，例如"2018-05-03"；date可以是None，当date=None时，返回最近一次更新的指数成份股权重。

    返回

        查询到对应日期，且有权重数据，返回 pandas.DataFrame， code(股票代码)，display_name(股票名称), date(日期), weight(权重)；
        查询到对应日期，且无权重数据， 返回距离查询日期最近日期的权重信息；
        找不到对应日期的权重信息， 返回距离查询日期最近日期的权重信息；
   
    '''

#获取行业板块成分股
def get_industry_stocks(industry_code, date=None) ->list:
    return LDClient.instance().get_industry_stocks(**locals())

    '''
    参数

        industry_code: 行业编码

        date: 查询日期, 一个字符串(格式类似’2015-10-15’)或者[datetime.date]/[datetime.datetime]对象, 可以是None, 使用默认日期. 这个默认日期在回测和研究模块上有点差别:

        回测模块: 默认值会随着回测日期变化而变化, 等于context.current_dt
        研究模块: 默认是今天

    返回

        返回股票代码的list
    '''


# 按照行业分类获取行业列表。
def get_industries(name, date=None) -> pd.DataFrame:
    return LDClient.instance().get_industries(**locals())
    '''
    # get_industries 获取行业列表
    from jqdata import *
    get_industries(name, date=None)

    按照行业分类获取行业列表。

    参数

        name: 行业代码， 取值如下：
            "sw_l1": 申万一级行业
            "sw_l2": 申万二级行业
            "sw_l3": 申万三级行业
            "jq_l1": 聚宽一级行业
            "jq_l2": 聚宽二级行业
            "zjw": 证监会行业

        date: 获取数据的日期，默认为None，返回历史上所有行业；传入date，返回date当天存在的行业；研究和回测中返回结果相同；

    返回值

        pandas.DataFrame， 各 column 的含义如下:
            index: 行业代码
            name: 行业名称
            start_date: 开始日期 

    示例

    from jqdata import *
    get_industries(name='zjw')
    get_industries(name='zjw', date='2016-01-01')
    '''

# 获取个股的所属行业数据
def get_industry(security, date=None) -> dict:
    return LDClient.instance().get_industry(**locals())
    '''
    参数
        security：标的代码。类型为字符串，形式如"000001.XSHE"；或为包含标的代码字符串的列表，形如["000001.XSHE", "000002.XSHE"]
        date：查询的日期。类型为字符串，形如"2018-06-01"或"2018-06-01 09:00:00"；或为datetime.datetime对象和datetime.date。注意传入对象的时分秒将被忽略。默认值为None，研究中默认值为当天，回测中默认值会随着回测日期变化而变化, 等于context.current_dt。
    返回结果
    一个dict， key是标的代码。
    示例
    >>> get_industry(security=['000001.XSHE','000002.XSHE'], date="2018-06-01")
    {'000001.XSHE': {'jq_l1': {'industry_code': 'HY007', 'industry_name': '金融指数'},
                    'jq_l2': {'industry_code': 'HY493', 'industry_name': '多元化银行指数'},
                    'sw_l1': {'industry_code': '801780', 'industry_name': '银行I'},
                    'sw_l2': {'industry_code': '801192', 'industry_name': '银行II'},
                    'sw_l3': {'industry_code': '851911', 'industry_name': '银行III'},
                    'zjw': {'industry_code': 'J66', 'industry_name': '货币金融服务'}
    参数

        security：标的代码，类型为字符串，形式如"000001.XSHE"；或为包含标的代码字符串的列表，形如["000001.XSHE", "000002.XSHE"]
        date：查询的日期。类型为字符串，形如"2018-06-01"或"2018-06-01 09:00:00"；或为datetime.datetime对象和datetime.date。注意传入对象的时分秒将被忽略。

    返回

    返回结果是一个dict，key是传入的股票代码

    示例

    #获取贵州茅台("600519.XSHG")的所属行业数据
    d = get_industry("600519.XSHG",date="2018-06-01")
    print(d)

    {'600519.XSHG': {'sw_l1': {'industry_code': '801120', 'industry_name': '食品饮料I'}, 'sw_l2': {'industry_code': '801123', 'industry_name': '饮料制造II'}, 'sw_l3': {'industry_code': '851231', 'industry_name': '白酒III'}, 'zjw': {'industry_code': 'C15', 'industry_name': '酒、饮料和精制茶制造业'}, 'jq_l2': {'industry_code': 'HY478', 'industry_name': '白酒与葡萄酒指数'}, 'jq_l1': {'industry_code': 'HY005', 'industry_name': '日常消费指数'}}}
    '''

# 获取所有标的信息
def get_all_securities(types=[], date=None) ->pd.DataFrame :
    return LDClient.instance().get_all_securities(**locals())
    '''
    获取平台支持的所有股票、基金、指数、期货、期权信息
    参数
        types: list: 用来过滤securities的类型, list元素可选: 'stock', 'fund', 'index', 'futures', 'options', 'etf', 'lof', 'fja', 'fjb', 'open_fund', 'bond_fund', 'stock_fund', 'QDII_fund'(QDII基金), 'money_market_fund', 'mixture_fund'。 types为空时返回所有股票, 不包括基金,指数和期货
        date: 日期, 一个字符串或者 datetime.datetime /datetime.date 对象, 用于获取某日期还在上市的股票信息. 默认值为 None, 表示获取所有日期的股票信息.建议使用时添加上指定date

    返回 [pandas.DataFrame], 比如:
    '''

# 获取单个标的信息

def get_security_info(code, date=None):
    return LDClient.instance().get_security_info(**locals())

    '''
        获取股票/基金/指数/期货的信息.
        参数
            code: 证券代码
            date：查询日期,默认为None，仅支持股票
        返回值
            一个对象, 有如下属性:
                display_name: 中文名称
                name: 缩写简称
                start_date: 上市日期, [datetime.date] 类型
                end_date: 退市日期（股票是最后一个交易日，不同于摘牌日期）， [datetime.date] 类型, 如果没有退市则为2200-01-01
                type: 股票、基金、金融期货、期货、债券基金、股票基金、QDII 基金、货币基金、混合基金、场外基金，'stock'/ 'fund' / 'index_futures' / 'futures' / 'etf'/'bond_fund' / 'stock_fund' / 'QDII_fund' / 'money_market_fund' / ‘mixture_fund' / 'open_fund'
                parent: 分级基金的母基金代码
        示例
        # 获取000001.XSHE的上市时间
        start_date = get_security_info('000001.XSHE').start_date
        print(start_date)
    '''


    '''
    参数
        security：标的代码。类型为字符串，形式如"000001.XSHE"；或为包含标的代码字符串的列表，形如["000001.XSHE", "000002.XSHE"]
        date：查询的日期。类型为字符串，形如"2018-06-01"或"2018-06-01 09:00:00"；或为datetime.datetime对象和datetime.date。注意传入对象的时分秒将被忽略。默认值为None，研究中默认值为当天，回测中默认值会随着回测日期变化而变化, 等于context.current_dt。
    返回结果
    一个dict， key是标的代码。
    示例
    >>> get_industry(security=['000001.XSHE','000002.XSHE'], date="2018-06-01")
    {'000001.XSHE': {'jq_l1': {'industry_code': 'HY007', 'industry_name': '金融指数'},
                    'jq_l2': {'industry_code': 'HY493', 'industry_name': '多元化银行指数'},
                    'sw_l1': {'industry_code': '801780', 'industry_name': '银行I'},
                    'sw_l2': {'industry_code': '801192', 'industry_name': '银行II'},
                    'sw_l3': {'industry_code': '851911', 'industry_name': '银行III'},
                    'zjw': {'industry_code': 'J66', 'industry_name': '货币金融服务'}
                    },
    '000002.XSHE': {'jq_l1': {'industry_code': 'HY011', 'industry_name': '房地产指数'},
                    'jq_l2': {'industry_code': 'HY509', 'industry_name': '房地产开发指数'},
                    'sw_l1': {'industry_code': '801180', 'industry_name': '房地产I'},
                    'sw_l2': {'industry_code': '801181', 'industry_name': '房地产开发II'},
                    'sw_l3': {'industry_code': '851811', 'industry_name': '房地产开发III'},
                    'zjw': {'industry_code': 'K70', 'industry_name': '房地产业'}
                    }
    }
    '''

def get_futures_info(securities=None, fields=('contract_multiplier','tick_size','trade_time')):
    return LDClient.instance().get_futures_info(**locals())
    
    '''
        infos = get_futures_info(["AG2012.XSGE",'A1609.XDCE' ])
        print(infos)
        >>>{'AG2012.XSGE': {'tick_size': 1.0, 
        'trade_time': [['2019-12-17', '2020-12-15', '21:00~02:30', '09:00~10:15', '10:30~11:30', '13:30~15:00']], 
        'contract_multiplier': 15.0}, 
        'A1609.XDCE': {'tick_size': 1.0, 
        'trade_time': [['2015-03-16', '2015-05-08', '21:00~02:30', '09:00~10:15', '10:30~11:30', '13:30~15:00'], 
        ['2015-05-09', '2016-09-14', '21:00~23:30', '09:00~10:15', '10:30~11:30', '13:30~15:00']],
        'contract_multiplier': 10.0}}
    '''

# get_all_trade_days 获取所有交易日
def get_all_trade_days():
    return LDClient.instance().get_all_trade_days(**locals())

    '''
    get_all_trade_days()

    获取所有交易日, 不需要传入参数, 返回一个包含所有交易日的 numpy.ndarray, 每个元素为一个datetime.date类型.

    注： 需导入 jqdata 模块，即在策略或研究起始位置加入

    '''
# get_trade_days 获取指定范围交易日

def get_trade_days(start_date=None, end_date=None, count=None):
    return LDClient.instance().get_trade_days(**locals())
    '''
    from jqdata import *
    get_trade_days(start_date=None, end_date=None, count=None)

    获取指定日期范围内的所有交易日, 返回一个包含datetime.date object的列表, 包含指定的 start_date 和 end_date, 默认返回至 datetime.date.today() 的所有交易日

    注意get_trade_days最多只能获取到截至现实时间的当前年份的最后一天的交易日数据

    注： 需导入 jqdata 模块，即在策略或研究起始位置加入
    import jqdata

    参数

        start_date: 开始日期, 与 count 二选一, 不可同时使用. str/datetime.date/datetime.datetime 对象
        end_date: 结束日期, str/datetime.date/datetime.datetime 对象, 默认为 datetime.date.today()
        count: 数量, 与 start_date 二选一, 不可同时使用, 必须大于 0. 表示取 end_date 往前的 count 个交易日，包含 end_date 当天。
    '''


